### ISSUE TITLE
Inappropriate assignment-from-none error

### ISSUE THREAD

**Issue by sam-s** [PyCQA/pylint#2559]
### Steps to reproduce
```python
class MyApp(metaclass=abc.ABCMeta):
    ...
    def lock_name(self):
        "Lock file name, mutex apps should redefine this to return a string."
        return None

    def lock_path(self):
        "Lock file path, based on lock_name()."
        lock_path = self.lock_name()
        if lock_path is None:
            return None
        ...
```

### Current behavior

```
 /.../application.py:109: refactor (R0201, no-self-use, MyApp.lock_name) Method could be a function
 /.../application.py:115: error (E1128, assignment-from-none, MyApp.lock_path) Assigning to function call which only returns None
```

### Expected behavior

The `no-self-use` message is, arguably, appropriate (although one could argue that since the class `MyApp` is _abstract_, the method _cannot_ be a function.

The message `assignment-from-none` is definitely wrong because _some_ descendants of `MyApp` _will_ redefine `lock_name` so that it will not _always_ return `None`.

### pylint --version output
```
pylint 2.1.1
astroid 2.0.4
Python 3.7.0 (default, Jun 29 2018, 20:13:13) 
[Clang 9.1.0 (clang-902.0.39.2)]
```

**Comment by PCManticore** [comment]
@sam-s thanks for creating an issue! I think a solution here would be to mark `lock_name` as an abstract method with `abc.abstractmethod`, as currently this check doesn't know if that method is supposed to be overridden in child classes.

(the `no-self-use` could be a separate issue)

**Comment by sam-s** [comment]
No, `lock_name` is _not_ an abstract method.
If it were abstract, I would have raised `NotImplementedError` there.
It is overridden by derived classes that want locking functionality, but not all derived classes want it.
Please reopen this issue.
Thanks.

**Comment by PCManticore** [comment]
@sam-s then there's no way for pylint to know that the method in question is supposed to return None or not, without a cue to indicate its status and return type. Given we don't look for annotations, it won't care as well if you annotate the method with `Optional[ExpectedType]`.

**Comment by sam-s** [comment]
I don't think you can actually infer that a _method_ **always** return `None` since it can always be overridden.

**Comment by earonesty** [comment]
In this setup, lock_path seems like an alias for lock_name.   Just use `lock_path = lock_name`.   You should never need to override things that way.

**Comment by sam-s** [comment]
> In this setup, lock_path seems like an alias for lock_name, there are better ways to do aliases.

no it is not.

**Comment by earonesty** [comment]
OK, I get it, the lock path function continues on afterward (...).   In general, abstract base classes should be exempt from the `no-self-use` and `unused-argument` rules anyway.  pylint just needs a way to know that a class was derived from or metaclasses as an ABC... and then skip some obviously incorrect rules.

**Comment by colindayechtra** [comment]
Here's a simpler case why this should not be an error.  The class `filter` is not abstract, it provides a meaningful default implementation that can be re-used by the application, and subclasses can be created to provided specialized filters.

```
class filter(object):
    def getFilter(self):
        return None  # default to "no filter"

class myFilter(filter):
    def getFilter(self):
        return "My special filter string"

# example use
defaultFilter = filter()
myFilter = myFilter()
processFiles(filter=defaultFilter)
processMoreFiles(filter=myFilter)

```

**Comment by rydrman** [comment]
I have also run into this issue... I completely agree that it's helpful to warn about the None return for a function with only empty return statements or no return statements at all. However when the function explicitly returns any kind of value it feels like an unsafe assumption that is will always be exactly that value - especially for subclasses (as we see in this case).

**Comment by Pierre-Sassoulas** [comment]
I'm reopening because of popular demand and because implementing abstract class exemption from the checks that do not make sense would eliminate a lot of false positives elsewhere too.
